### PR TITLE
Fix #765 & #763 & #768

### LINKED ISSUE #763: `string_lit_as_bytes` gives incorrect recommendation with stringify!()
Sample program:

```rust
#![feature(plugin)]
#![plugin(clippy)]

fn main() {
    let _ = bstringify!(foo).as_bytes();
}
```

Warning given:
> calling `as_bytes()` on a string literal. Consider using a byte string literal instead: `bstringify!(foo)`, #[warn(string_lit_as_bytes)] on by default

Of course, using `bstringify!(foo)` gives an undefined macro error from rustc.

### LINKED ISSUE #765: thread 'rustc' panicked at 'arithmetic operation overflowed'
Clippy 0.0.51 causes a panic when compiled on Travis:

https://travis-ci.org/badboy/iso8601/jobs/115904545

Seems it is this line: https://github.com/Manishearth/rust-clippy/blob/4fa6f49f74f9a60ec4df1a46d58e49d8e5838432/src/array_indexing.rs#L135

I can't reproduce it on my machine at the moment.

### PR DESCRIPTION
**Pull request by mcarton**
Fix #765. Fix  #763. Ref #442.

### CODE REVIEW

File: src/strings.rs [hunk 1]
```diff
@@ -140,12 +140,19 @@ impl LateLintPass for StringLitAsBytes {
...
-                                              snippet(cx, args[0].span, r#""foo""#));
-                            span_lint(cx, STRING_LIT_AS_BYTES, e.span, &msg);
+                        if lit_content.chars().all(|c| c.is_ascii()) && !in_macro(cx, args[0].span) {
+                            span_lint_and_then(cx,
+                                               STRING_LIT_AS_BYTES,
+                                               e.span,
+                                               "calling `as_bytes()` on a string literal",
+                                               |db| {
+                                                   let sugg = format!("b{}",
+                                                                      snippet(cx, args[0].span, r#""foo""#));
```
Comment (Manishearth) [inline src/strings.rs]: this should be e.span, perhaps?

**Comment by Manishearth** [comment]
We have a merge conflict here with the const changes (nontrivial)

**Comment by oli-obk** [comment]
i can fix it tomorrow

**Comment by mcarton** [comment]
Wasn’t that bad.

**Comment by mcarton** [comment]
Also fix #768.

**Pull request merged**